libcaf_core/src/detail/test_actor_clock.cpp

```diff
@@ -46,10 +50,34 @@ size_t test_actor_clock::trigger_timeouts() {
 }
 
 size_t test_actor_clock::advance_time(duration_type x) {
-  CAF_LOG_TRACE(CAF_ARG(x) << CAF_ARG2("schedule.size", schedule_.size()));
+  CAF_LOG_TRACE(CAF_ARG(x) << CAF_ARG2("schedule.size", schedule.size()));
   CAF_ASSERT(x.count() >= 0);
   current_time += x;
-  return trigger_expired_timeouts();
+  auto result = size_t{0};
+  while (!schedule.empty() && schedule.begin()->first <= current_time)
+    if (try_trigger_once())
+      ++result;
+  return result;
+}
+
+bool test_actor_clock::try_trigger_once() {
+  auto i = schedule.begin();
+  auto t = i->first;
+  if (t > current_time)
+    return false;
+  auto [f, period] = i->second;
+  schedule.erase(i);
+  if (f.run() == action::transition::success) {
+    if (period.count() > 0) {
+      auto next = t + period;
+      while (next <= current_time)
+        next += period;
+      schedule.emplace(next, schedule_entry{std::move(f), period});
+    }
+    return true;
+  } else {
+    return false;
+  }
```

Comment:
An action cannot *fail* in the sense that it'd be an error. It may get *disposed* asynchronously, in which case running it becomes a no-op. Whether the action actually did something or not is irrelevant to the clock here.